### PR TITLE
TYPE: Add a StatementUpDownMover for function

### PR DESCRIPTION
**Pull request by farodin91**
Todo

- [x] function
- [x] struct
- [x] tests 
- [x] edge case
- [x] move documentation
- [x] more generic test much duplicated code
- [ ] more generic for example mod, impl, block, trait, if, loops, match pattern, macro pattern, macros, ...

### CODE REVIEW

**Comment by farodin91** [comment]
Fixes #1459

File: src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        """, """
+fn test() {
+
+}
+fn foo() {}
+        """, false)
+
+    fun moveTest(
+        @Language("Rust") before: String,
+        @Language("Rust") after: String, down: Boolean) {
```
Comment (matklad) [inline src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt]: Boolean parameters like `down: Boolean` often are code smell. Let's make two functions instead, `testMoveUp` and `testMoveDown`, to make it easier to find out what is supposed to happen at a glance!

**Comment by matklad** [comment]
Looks nice! It looks like the logic here could grow very complex, so it would be helpful to have more tests here!

**Comment by farodin91** [comment]
How to run format or ignore ident after an action? I didn't like the current test compare.

**Comment by farodin91** [comment]
@matklad I will start refactor the code. Now I have unit tests.

**Comment by farodin91** [comment]
Some refactor done.

**Comment by farodin91** [comment]
@matklad Do you like to review the current state.

File: src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import com.intellij.openapi.editor.Editor
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.ide.utils.findElementAtIgnoreWhitespaceAfter
+import org.rust.ide.utils.findElementAtIgnoreWhitespaceBefore
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsMod
+import org.rust.lang.core.psi.ext.parentOfType
+
+val PsiElement.line: Int? get() = containingFile.viewProvider.document?.getLineNumber(textRange.startOffset)
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt]: This probably wants to be private?

File: src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+abstract class RsAbstractMover : StatementUpDownMover() {
+
+    abstract val listOfContainers: List<Class<out PsiElement>>
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt]: Let's name it just `containers`. Usually it is not necessary to repeat the name of the type in the name of variable

File: src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            info.toMove.endLine + 1
+        }
+        if (line < 0 || line >= editor.document.lineCount) {
+            return info.prohibitMove()
+        }
+        val offset = editor.document.getLineStartOffset(line)
+        var element = if (!down) {
+            file.findElementAtIgnoreWhitespaceBefore(offset) ?: return false
+        } else {
+            file.findElementAtIgnoreWhitespaceAfter(offset) ?: return false
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt]: `?: return false` can be pulled out of the `if` expression.

File: src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        while (element != null && element !is RsFile) {
+            if (listOfContainers.any { it.isAssignableFrom(element.javaClass) }) {
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt]: This reflection does not look exactly pretty...

What do you think about using `IElementTypes` instead of classes (like `STRUCT_ITEM`). Alternatively, subclasses can provide a method like `isContainer(psi: PsiElement)`, and use a `when` inside.

**Comment by farodin91** [comment]
Other `mover` could be also handled in a different PR.

File: src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+                        return info.prohibitMove()
+                    }
+                    parentOfFn = parentOfFn.parent
+                }
+            }
+
+            if (jumpOver.any { it.isAssignableFrom(element.javaClass) }) {
+                break
+            }
+            val context = element.context ?: break
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt]: Up/Down always operates with real files, so there's no need to use `context` here, `parent` is sufficient.

File: src/test/kotlin/org/rust/ide/actions/RsStructItemMoverTest.kt [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+        """, """
+            // - main.rs
+            fn s() {
+                struct /*caret*/ A{
+                    test: u32
+                }
+            }
+        """)
+
+    fun moveUpTest(@Language("Rust") before: String, @Language("Rust") after: String) {
```
Comment (matklad) [inline src/test/kotlin/org/rust/ide/actions/RsStructItemMoverTest.kt]: These functions could be `private`.

**Comment by farodin91** [comment]
Resolved

**Comment by farodin91** [comment]
@matklad Resolved merge conflict. Any thing to do?

**Comment by farodin91** [comment]
I found a bug.

File: src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+            // - main.rs
+            fn foo() {
+
+            }
+            fn /*caret*/test() {
+
+            }
+        """)
+
+    // This test fails in run intellij
```
Comment (farodin91) [inline src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt]: @matklad Any suggestions why this happened?

Comment (matklad) [inline src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt]: No, of the top of my head I don't know what happens here. So looks like the best way to debug this is to launch IDEA under debugger and just step through. 

However, the behavior in IntelliJ seems more reasonable: if there's a single blank line between two functions, we should swap them. Kotlin preserves all the whitespace between functions as well, so perhaps we should do the same?

File: src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+import com.intellij.psi.PsiFile
+import com.intellij.psi.tree.IElementType
+import org.rust.ide.utils.findElementAtIgnoreWhitespaceAfter
+import org.rust.ide.utils.findElementAtIgnoreWhitespaceBefore
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.elementType
+
+val PsiElement.line: Int? get() = containingFile.viewProvider.document?.getLineNumber(textRange.startOffset)
+
+abstract class RsAbstractMover : StatementUpDownMover() {
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RsAbstractMover.kt]: Let's name it `RsAbstractUpDownMover`, to make it easier to discover via <kbd>Ctrl+N</kbd>? The descendants should probably have `UpDown` in their names as well.

File: src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt (hunk 8 above)
Comment (farodin91) [inline src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt]: I known, but test passes.

Comment (matklad) [inline src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt]: Yeah, that's a complete mystery to me as well!

Unfortunately, I have no better hints then step through in test and in live IDEA and to try to find out, what's different.

Comment (farodin91) [inline src/test/kotlin/org/rust/ide/actions/RsFunctionMoverTest.kt]: I find the problem space are different handle to new lines.

**Comment by matklad** [comment]
For this code

```
struct Foo {
    foo: u32,
}

impl Foo {}
```

moving the `impl` up won't jump over `struct`, but it should I think.

**Comment by farodin91** [comment]
I only support currently structs and function.

**Comment by matklad** [comment]
Ok, it can be handled in a separate PR!

Thanks!

bors r+

**Comment by farodin91** [comment]
bors failed?
